src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs

```diff
@@ -673,27 +673,28 @@ private BoundStatement BindDeclarationStatementParts(LocalDeclarationStatementSy
             originalBinder.BindForOrUsingOrFixedDeclarations(declarationSyntax, LocalDeclarationKind.UsingVariable, diagnostics, out declarations);
             Debug.Assert(!declarations.IsEmpty);
 
-            BoundTypeExpression declTypeExpr = declarations[0].DeclaredType;
+            BoundLocalDeclaration declaration = declarations[0];
+            TypeSymbol declType = declaration.DeclaredType.Type;
 
-            if (declTypeExpr.Type.IsDynamic())
+            if (declType.IsDynamic())
             {
                 iDisposableConversion = Conversion.ImplicitDynamic;
             }
             else
             {
                 HashSet<DiagnosticInfo> useSiteDiagnostics = null;
                 TypeSymbol iDisposable = this.Compilation.GetSpecialType(SpecialType.System_IDisposable);
-                iDisposableConversion = originalBinder.Conversions.ClassifyImplicitConversionFromType(declTypeExpr.Type, iDisposable, ref useSiteDiagnostics);
+                iDisposableConversion = originalBinder.Conversions.ClassifyImplicitConversionFromType(declType, iDisposable, ref useSiteDiagnostics);
                 diagnostics.Add(declarationSyntax, useSiteDiagnostics);
 
                 if (!iDisposableConversion.IsImplicit)
                 {
-                   disposeMethod = TryFindDisposePatternMethod(declTypeExpr, declarationSyntax, diagnostics);
+                   disposeMethod = TryFindDisposePatternMethod(declaration.InitializerOpt, declarationSyntax, diagnostics);
```

Comment:
>declaration.InitializerOpt [](start = 63, length = 26)

I'm surprised that we look at the initializer expression, rather than the declared type.
`using (TypeWithDipose x = new TypeWithoutDispose()) ...` (with a user-defined conversion from one type to another) would be a relevant scenario.

Reply:
Fixed + added tests to cover.